BUG: GH13219 Fixed. Allow unicode values in usecol

hassanshamim opened the pull request:
- [x] closes #13219 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

This is my first time contributing to this project.  Let me know what needs to be improved, especially with regards to the test and the bugfix.  It was simpler than I imagined which makes me want to say I did it wrong.

hassanshamim committed:
BUG: GH13219 Fixed. Allow unicode values in usecol

sinhrks commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,25 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+        }
+        expected = DataFrame(data)
+
+        df = self.read_csv(StringIO(s), usecols=[u'AAA', u'BBB'])
```
Can you also test mixed str, like ``usecols=[u'AAA', 'BBB']``

hassanshamim replied:
Good idea.  Will update.

hassanshamim replied:
Should mix strings like `usecols=[u'AAA', 'BBB']` be successful?  Or should it throw out a more descriptive error, like `ValueError: The elements of 'usecols' must all be of the same type`

jreback replied:
that already should raise because

```
In [2]: pd.lib.infer_dtype([u'AA', 'BB'])
Out[2]: 'mixed'
```
(certainly add a test for it)

sinhrks replied:
Ideally it should success. It is quite popular in countries using 2bytes. I don't think checking all dtypes in "mixed" can be a bottleneck.

hassanshamim replied:
What's the approach then?  Recode unicode values to UTF8 if the dtype is mixed in the [_validate_usecols_arg](https://github.com/pydata/pandas/blob/c9ffd7891dadd6e5590695e142f77a3476b5c4e3/pandas/io/parsers.py#L843)? or just accept 'mixed' type if they are all string_type?

sinhrks replied:
I thought something like:

```
if usecols_dtype == "mixed" and not all(isinstance(x, compat.string_types)):
    raise...
```

Because ``usecols`` length is less likely to be too long.

gfyoung replied:
@jreback : in reference to what you said <a href="https://github.com/pydata/pandas/pull/13233#discussion_r64140733">here</a>, then what @sinhrks is suggesting above should not have been implemented at all.

jreback replied:
This is working as expected, so something else is going on.

```
In [4]: pd.lib.infer_dtype(u'ああ,いい,ううう,ええええ'.split(','))
Out[4]: 'unicode'

In [5]: pd.lib.infer_dtype([u'A',u'B'])
Out[5]: 'unicode'

In [6]: pd.lib.infer_dtype([u'A','B'])
Out[6]: 'mixed'

In [7]: pd.lib.infer_dtype(['A','B'])
Out[7]: 'string'
```

gfyoung replied:
@jreback : I'm confused...what exactly were you checking here?  Not sure if it is related to this discussion here with testing with mixed string-type `usecols` (see first comment by @sinhrks )

jreback replied:
look at the code, this is EXACTLY what it does. it already check ``string`` and ``integer``, just add ``unicode`` and it should work. if it doesn't (and I guess the multi-byte is failing it), then there is something else to look at that is causing a failure. 

@hassanshamim pls step thru and see where it fails. This PR is getting way overcommented with a really simple thing.

sinhrks commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,25 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+    def test_usecols_accepts_unicode_strings(self):
+        # see gh-13219
+
+        s = u'''AAA,BBB,CCC,DDD
```
Can u also add a case with 2byte characters, like:

```
あああ,いい,ううう,ええええ
0.056674973,8,True,a
2.613230982,2,False,b
3.568935038,7,False,a
```

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -154,3 +154,4 @@ Bug Fixes
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``Peirod`` and ``Series`` or ``Index`` comparison raises ``TypeError`` (:issue:`13200`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Bug in ``io.parsers.read_csv()`` that prevents ``usecol`` kwarg from accepting unicode (:issue:`13219`)
```
``pd.read_csv()``

hassanshamim commented:
Currently the multibyte usecol tests are throwing errors for the test parsers.  Is this expected?  If so I'll adjust the tests.

```

======================================================================
ERROR: test_usecols_with_multibyte_unicode_characters (pandas.io.tests.parser.test_parsers.TestCParserHighMemory)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/usecols.py", line 309, in test_usecols_with_multibyte_unicode_characters
    df = self.read_csv(StringIO(s), usecols=[u'あああ', u'いい'])
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/test_parsers.py", line 57, in read_csv
    return read_csv(*args, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 562, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 315, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 645, in __init__
    self._make_engine(self.engine)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 799, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 1257, in __init__
    raise ValueError("Usecols do not match names.")
ValueError: Usecols do not match names.

======================================================================
ERROR: test_usecols_with_multibyte_unicode_characters (pandas.io.tests.parser.test_parsers.TestCParserLowMemory)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/usecols.py", line 309, in test_usecols_with_multibyte_unicode_characters
    df = self.read_csv(StringIO(s), usecols=[u'あああ', u'いい'])
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/test_parsers.py", line 76, in read_csv
    return read_csv(*args, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 562, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 315, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 645, in __init__
    self._make_engine(self.engine)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 799, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 1257, in __init__
    raise ValueError("Usecols do not match names.")
ValueError: Usecols do not match names.

======================================================================
ERROR: test_usecols_with_multibyte_unicode_characters (pandas.io.tests.parser.test_parsers.TestPythonParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/usecols.py", line 309, in test_usecols_with_multibyte_unicode_characters
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/tests/parser/test_parsers.py", line 100, in read_csv
    return read_csv(*args, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 562, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 315, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 645, in __init__
    self._make_engine(self.engine)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 805, in _make_engine
    self._engine = klass(self.f, **self.options)
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 1608, in __init__
    self.columns, self.num_original_columns = self._infer_columns()
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 1823, in _infer_columns
    line = self._buffered_line()
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 1975, in _buffered_line
    return self._next_line()
  File "/Users/avendesora/Dropbox/code/python/pandas-hassan/pandas/io/parsers.py", line 2006, in _next_line
    orig_line = next(self.data)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-2: ordinal not in range(128)

----------------------------------------------------------------------
Ran 587 tests in 111.477s

FAILED (SKIP=21, errors=3)

```

sinhrks commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -154,3 +154,4 @@ Bug Fixes
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``Peirod`` and ``Series`` or ``Index`` comparison raises ``TypeError`` (:issue:`13200`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Bug in ``pd.read_csv()`` that prevents ``usecol`` kwarg from accepting unicode (:issue:`13219`)
```
So we can't say "accepting unicode" in a broad sense, as there is limitation for 2bytes now.

sinhrks commented:
This _should_ work, but currently not. Can we keep open #13219 to handle 2 bytes?

jreback commented:
how about we comment out these 2 byte tests for now (and make a new issue)?

@hassanshamim can you do that & rebase?

hassanshamim commented:
Can do!

jreback commented on pandas/io/parsers.py:
```diff
@@ -847,12 +847,18 @@ def _validate_usecols_arg(usecols):
-                              "must either be all strings "
-                              "or all integers"))
+
+        if usecols_dtype == 'mixed':
```
I don't think this is necessary, only the bottom check.

hassanshamim replied:
Do you mean the  `usecols_dtype not in ('integer', 'string', 'unicode'):`?  I  took @sinhrks advice and by checking if the type is mixed and ensuring they are all string_type, it allows usecols of mixed encoding strings, and `test_usecols_with_mixed_encoding_strings` passes.

Or if you mean just have the `if not all(map(lambda x: isinstance(x, string_types), usecols))` line, then just having that would fail with mixed integer and strings, as those are `mixed-integer` dtype rather than just `mixed`.

jreback replied:
only valid types are string, integer, Unicode 

mixed is not allowed (your check is duplicative)

gfyoung replied:
@jreback : that's in conflict with what @sinhrks is suggesting <a href="https://github.com/pydata/pandas/pull/13233/files/71f5139f14b38a1a3c1d9fd7d9219947399faec0#r63986201">below</a>.  Perhaps that should be sorted out?

jreback replied:
i c - I think then that the infer types is failing on the multi byte Unicode check (which is what should be fixed)

gfyoung replied:
@jreback : What about a new `lib` method like `is_string_unicode_array` (similar to `is_integer_float_array`?  I don't see anything wrong with the `dtype` inference checks for `string` and `unicode` at the moment.

@Everyone: For anyone who doesn't quite know / remember where the source code for `infer_dtype` is, you can find it <a href="https://github.com/pydata/pandas/blob/master/pandas/src/inference.pyx#L83">here</a>.

jreback replied:
I think it should work now

can u directly test the inference function with multi byte?

gfyoung replied:
@jreback : What did you do exactly?

jreback replied:
what I mean is it looks like it should work
if it doesn't something weird is going on

gfyoung replied:
Well I can't comment on the errors that @hassanshamim is getting when running the multibyte, but I was suggesting that there should be a nice helper function to check for a mixed array of `unicode` and `string` (we might be talking about different things here?)

jreback replied:
mixed strings and Unicode are an error full stop

gfyoung replied:
@jreback : that's where @sinhrks disagrees, hence my comment previously.  I thought you had understood that when said "i c"?

jreback replied:
no strings are either Unicode or not

hassanshamim replied:
So should I remove the first check and revert to the original, where passing mixed string and unicode i.e. `usecols=[u'AA', 'BB']` fails?

jreback commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,67 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+        df = self.read_csv(StringIO(s), usecols=[u'AAA', 'BBB'])
+        tm.assert_frame_equal(df, expected)
+
+#    def test_usecols_with_multibyte_unicode_characters(self):
```
put a TODO here. (in a comment) as well as create a new issue and put the reference here

gfyoung replied:
To piggy-back on @jreback's comment: write a test for single-byte unicode characters as well.

gfyoung replied:
In addition, can you include a test where the `usecols` in that `multibyte` test that you wrote aren't explicitly `unicode`?  I don't get any errors raised for either engine using `0.18.1`.

hassanshamim replied:
@gfyoung by 'single-byte unicode characters' do you mean ASCII chars in unicode?  i.e. `u'A'` (or rather `pd.compat.u('A')`? Or just any unicode char whose codepoint starts with 00? This python2 vs python3 string/unicode thing is starting to make my head hurt and just want to make sure I understand.  Example please 😅

gfyoung replied:
@hassanshamim : `u'A'`.  It's just for completeness.  Sorry for the confusion!

gfyoung commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,67 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+        # see gh-13219
+
+        s = u'''AAA,BBB,CCC,DDD
+        0.056674973,8,True,a
```
Don't make the data string unicode (all other `parser` tests don't).  You are just testing that `read_csv` accepts `usecols` with `unicode` strings.  If you want to include a test where `unicode` data is mixed with `unicode` and non-`unicode` strings, write that as a separate test from this one.

gfyoung commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,67 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+
+    def test_usecols_with_mixed_encoding_strings(self):
+        s = u'''AAA,BBB,CCC,DDD
+        0.056674973,8,True,a
```
Ditto regarding the `unicode` data.

hassanshamim commented:
[x] Created Issue #13253
[x] Converted data strings from unicode to str in tests
[x] Added test for multibyte characters (not unicode)
[x] Reverted change that allowed `mixed` dtype usecol values
[x] Rebased onto upstream/master

I think that's everything.  I've been squashing my new commits and just force pushing to update this branch.  Is that the normal process for incorporating feedback on PRs?  Or should I have just left multiple commits then squashed once this was all ready?

gfyoung commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,76 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+                          usecols=['AAA', u'BBB'])
+        self.assertRaises(ValueError, self.read_csv, StringIO(s),
+                          usecols=[u'AAA', 'BBB'])
+
```
Use `assertRaisesRegexp` from `pandas.util.testing` to test that you get the right error message.  That's stronger than `assertRaises`.

gfyoung commented:
@hassanshamim : 
1) You forgot to add test for single-byte `unicode` `usecols` (i.e. `u'A'`)
2) Regarding squashing, I think that in this case, it is okay to squash because this really should be one commit.  In cases where there might be a lot more changing, multiple commits would be preferred.

jreback commented:
this looks good. @gfyoung any further comments?

gfyoung commented on pandas/io/parsers.py:
```diff
@@ -882,12 +882,13 @@ def _validate_usecols_arg(usecols):
     """
+    msg = ("The elements of 'usecols' must "
+           "either be all strings or all integers")
+
```
If we're going to raise on a mixture of `unicode` and `string`, the current error message is not very useful in that case.   Either add to this error to account for that case, OR have a different error message.

gfyoung commented on pandas/io/tests/parser/usecols.py:
```diff
@@ -254,3 +254,103 @@ def test_usecols_with_parse_dates_and_usecol_names(self):
+
+#    TODO: See gh-13253
+#    def test_usecols_with_multibyte_unicode_characters(self):
+#       s = '''あああ,いい,ううう,ええええ
```
Use `raise nose.Skiptest(...)` instead of commenting this whole test out

gfyoung commented:
@hassanshamim : just a few comments, but almost there!

hassanshamim commented:
@gfyoung updated with your recommendations.  Are these formats okay?  
````
def test_usecols_with_multibyte_unicode_characters(self):
    raise nose.SkipTest('TODO: see gh-13253')
    # actual test code...
````

New error message:
````
msg = ("The elements of 'usecols' must "
            "either be all strings, all unicode, or all integers")
````

gfyoung commented:
@hassanshamim : yep, that looks good, so LGTM now.  Just ping when tests pass!

hassanshamim commented:
@gfyoung tests have gone through.

jreback commented:
thanks!
